Zone cleaning functionality implemented for the PUREi9

joakimjalden opened the pull request:
## What

### Code

The PR implements the `clean_zones` command for the PUREi9 RVCs. The base syntax of the command and parameters is compatible with the syntax of [home-assistant-purei9](https://github.com/Ekman/home-assistant-pure-i9/tree/master). That said, I wanted to provide the full capabilities, which for the PUREi9 also includes setting the `power_mode`. Hence, a `zones` zone can be either specified by only the name of the zone or with a dict that contains the `name` and optionally the `power_mode` of the zone. If no power mode is provided, the default specified in the map is used.

The code utilizes `voluptuous` (as the rest of HA) to validate the user-supplied parameters. It reports back errors using the `ServiceValidationError` as specified in the [HA developers' documentation](https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/action-exceptions/). This provides user-interpretable error messages for syntax errors in the GUI. I've done extensive testing and hope that I covered even the corner cases with this.

### Manual

Since the usage of the command is non-trivial, I created a small manual alongside `README.md` to explain its usage. It is listed after the device list in `README.md` (where I also removed a duplicate entry). Please let me know if you would like this information elsewhere, or feel free to modify it as needed. 

## Caveats and discussions

Several design choices were not immediately apparent, and I could use a second opinion.

### Power mode naming

The power modes are now set using the `power_mode` parameter. The options are `[Quiet|Smart|Power]` (strings with the first letter capitalized). The syntax for the power modes is taken from the [Electrolux developer docs](https://developer.electrolux.one/documentation/reference#getMemoryMaps). That said, I am not entirely satisfied with the capitalization when used as a parameter. The Electrolux API for the PUREi9 takes an integer for the power mode, so I was considering making it an integer in HA as well. Still, since the 700 series takes a `vacuumMode` parameter with options `[quiet| energySaving|standard|powerful]` (supplied as strings to the Electrolux API, with camelCase), we will eventually have to convert from strings to numbers anyway if we wish consistency in HA.

I see at least three reasonable options for the naming convention.
1. As it is now, with capitalized strings following the Electrolux API conventions.
2. Making all strings that the user sees lowercase (or possibly snake case when it comes to the 700 series)
3. Making the case checking logic case agnostic and letting the user decide what looks best in the YAML
I understand that it is pedantic and subjective, but any input would be appreciated. It would be beneficial to settle on this before incorporating the code, as future changes to this may break already created actions.

I was also considering naming the option simply `mode` instead of `power_mode` to create some similarity with the `vacuumMode` option for the 700 series vacuums, as discussed below.

### 700 series implementation

The code only implements the functionality for the PUREi9 RVC since I do not have a 700 series RVC to test with. That said, I have been reviewing the 700 series documentation to ensure the (hopefully) upcoming implementation is as smooth as possible. Although it does not directly relate to this PR, it is beneficial to look ahead.

I see two different ways of implementing the corresponding functionality for the 700 series.

1. Recognizing that the capabilities of the 700 are different, and implementing a new command `clean_rooms` that more closely corresponds to those, acknowledging the differing terminology (`rooms` versus `zones`), and the greatly expanded set of options per room. It may also be more intuitive for some users if the Wellbeing app uses different terminology.

2. Expand `clean_zones` by adding additional options to the `zones` params schema to map to RVC functionality, such as `vacuumMode`, `sweepMode`, `waterPumpRate`, and `numberOfCleaningRepetitions`. The fact that Electrolux now refers to the zones as rooms would be hidden from the HA user, and `vacuumMode` and `power_mode` could potentially be combined into a single `mode` option.

Input is appreciated. Once the current PR is handled, I can create a local branch with the 700 series functionality to guide implementation, which can then be tested once an owner of such an appliance requests it. Since these vacuums are not handled by the home-assistant-purei9 integration, it would be a unique feature.

joakimjalden commented:
Upon further reflection and examination of the Electrolux app, I believe it would be more intuitive to use `mode` in place of `power_mode` and to use strings with capitalized letters, as this aligns with the naming convention in the app. I went ahead and changed `power_mode` to `mode` in the PR. When dealing with the 700 series, we should also follow the naming conventions of the Electrolux app for the user-facing strings in HA.

![purei9_interface](https://github.com/user-attachments/assets/a61b959c-6ec9-4a01-bbfc-658a9e1dc04a)

JohNan commented:
I suggest that you make the mode case agnostic and/or snake case, since it looks like HA mostly use snake case in their yaml. 

Naming it `mode` sounds good since it's what the app says. Consistency and intuitiveness is good. So I also suggest that we go with the first option regarding the 700 series. 

If the vacuums has different params I think it's better to keep it separated so that the user know what params to use when.

joakimjalden commented:
I'm sorry for adding so many (unrequested) updates to this PR. However, after reading up on the FAN_SPEED feature of the vacuum integration, I realized that one of the effects of enabling this (later) will be the addition of the action vacuum.set_fan_speed, where an option (data attribute) will be `fan_speed`. Since this will have the same function (globally) and options as the `mode` I introduced for `zone_cleaning`, I changed the name `mode` to `fan_speed` for consistency across the full integration. I also thought that `zone` was a better convention than just `name` (which is a bit non-descript). I thought it was better to address this now before anyone else uses the functionality.

joakimjalden commented:
I addressed the copilot's comments regarding changing the name of the fan speeds constant, and breaking out and commenting the lambda function in the Schema validation and commented the functionality. I did not add test cases for `clean_zones` as this would require adding test infrastructure that is not yet available, and I did not write a helper for similar ServiceErrors, as this code is similar in style to error logging elsewhere (and would require a more uniform change).

I have retested and confirmed that everything works on my system.

It should be ready for merging.
